### PR TITLE
Feature deriv

### PR DESCRIPTION
**Pull request by tj-sun**
This pull request creates Derivative class

- Create Derivative class for better handling of PDE
- Improved algorithm to calculate bc (re calculating normal stresses), error much smaller now
- Move helper functions to util.py
- improved bc calculation, bc is passed as symbolic equations instead of strings
- improved reading media parameters from file, handled in StaggeredGrid now, more robust algorithm to replace with effective media parameters
- updated ipython notebooks
- eigenwave test default to source code generation only, need --compile flag to trigger compilation
- updated travis.yml scrip to reflect above

### CODE REVIEW

File: opesci/fields.py [hunk 1]
```diff
@@ -359,13 +167,15 @@ def set_dt(self, dt):
...
-        - these parameters should be used for the updating kernel
-        - only needed when read data from file
+        link this velocity field to a list of stress field
+        e.g. Vx will be associated with Txx, Txy, Txz
+        link compression stress field with other compression stress field
+        e.g. Txx with Txx, Tyy, Tzz
+        :param sfields: list of associated stress field
         """
-        self.media_param = media_param
+        self.sfields = sfields
```
Comment (mlange05) [inline opesci/fields.py]: Shouldn't this method be specific to `VField` objects, rather than `Field`?

File: opesci/staggeredgrid.py [hunk 2]
```diff
@@ -903,24 +990,36 @@ def stress_bc(self):
...
+                                else:
+                                    body = ''.join(ccode_eq(bc)+';\n' for bc in field.bc[d+1][side])
                                 dict1 = {'i': i, 'i0': i0,
                                          'i1': i1, 'body': body}
-                                body = render(tmpl, dict1).replace('[t]',
-                                                                   '[t1]')
+                                body = render(tmpl, dict1).replace('[t + 1]', '[t1]').replace('[t]', '[t0]')
                                 if self.ivdep:
-                                    body = '%s\n' % self.compiler._ivdep + body
+                                    body = '#pragma ivdep\n' + body
```
Comment (mlange05) [inline opesci/staggeredgrid.py]: The `ivdep` pragma is compiler-dependent, ie. GCC requires it to be `#pragma GCC ivdep`, so we should get this from the compiler object.

File: tests/eigenwave3d.py [hunk 3]
```diff
@@ -247,9 +240,9 @@ def main():
 """
     p = ArgumentParser(description="Standalone testing script for the Eigenwave3D example",
                        formatter_class=RawTextHelpFormatter)
-    p.add_argument('mode', choices=('default', 'read', 'converge', 'cx1'),
+    p.add_argument('mode', choices=('default', 'read', 'converge', 'cx1', 'source'),
```
Comment (mlange05) [inline tests/eigenwave3d.py]: I don't think we need the new mode anymore.

File: opesci/fields.py [hunk 4]
```diff
@@ -391,53 +201,61 @@ def set(self, dimension, direction):
...
+        derivatives = get_all_objects(expr, DDerivative)
+        for deriv in derivatives:
+            # using 2nd order approximation
+            dict1[deriv] = deriv.fd[1]
+            # if deriv.var == idx[self.direction]:
+            #     # e.g. using 2nd order approximation for dW/dz at z surface
+            #     dict1[deriv] = deriv.fd[1]
+            # else:
+            #     # e.g. using 4th order approximation for dU/dz at z surface
+            #     dict1[deriv] = deriv.fd[2]
```
Comment (mlange05) [inline opesci/fields.py]: Please remove dead code.

File: opesci/fields.py [hunk 5]
```diff
@@ -391,53 +201,61 @@ def set(self, dimension, direction):
...
-        if read:
-            rhs = rhs.subs(self.media_param)
-            rhs = rhs.subs(indices[d], b)
+        lhs = lhs.subs(self.indices[d], t)
+        rhs = self.align(rhs.subs(self.indices[d], t))
+        # # if read data from file, replace media parameters with array
+        # # replace particular index with boundary
+        # if read:
+        #     rhs = rhs.subs(self.media_param)
+        #     rhs = rhs.subs(self.indices[d], b)
```
Comment (mlange05) [inline opesci/fields.py]: Please remove dead code.

**Comment by mlange05** [comment]
I think the new defaults require a small README update, but apart from that and the in-line comments, this looks very good to me.

File: opesci/fields.py (hunk 1 above)
Comment (tj-sun) [inline opesci/fields.py]: stress field also need to be associated with another stress field for bc.
e.g. Txx need Txz to update at the z boundary

File: tests/eigenwave3d.py (hunk 3 above)
Comment (tj-sun) [inline tests/eigenwave3d.py]: yeah that's right. deleted now.

File: opesci/fields.py (hunk 4 above)
Comment (tj-sun) [inline opesci/fields.py]: deleted now.

File: opesci/fields.py (hunk 5 above)
Comment (tj-sun) [inline opesci/fields.py]: deleted now.

**Comment by mlange05** [comment]
Looks great now. Merging...

**Pull request merged**